### PR TITLE
GEM and ME0 segments into trackerMuon

### PR DESCRIPTION
**Pull request by jshlee**
This PR adds GemMuon and Me0Muon as well as adding GEMSegments and ME0Segments into trackerMuons.

TAMuonSegmentMatch and TrackAssociatorParameters
- added GEM segments
- added ME0 segments

MuonIdProducer 
- added isGEMMuon function
- added isME0Muon function
- save isME0Muon and isGEMMuon with trackerMuons

muon.h
- added numberOfChambersNoGEM function
- added numberOfChambersNoME0 function
- added isGEMMuon function
- added isME0Muon function

MuonChamberMatch, MuonDetIdAssociator, DetIdInfo, TAMuonChamberMatch, TrackDetectorAssociator
- added ME0 and GEM 

ME0SegmentBuilder
- fix chamberId()

MuonME0DetLayerGeometryBuilder
- bug fix, no layer 0

### CODE REVIEW

**Comment by jshlee** [comment]
@calabria and @kpedro88 - PR for GEMMuon and ME0Muon

**Comment by kpedro88** [comment]
@jshlee - reminder: at some point, we need a long-term fix for MuonSegFit:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/RecoLocalMuon/GEMSegment/plugins/MuonSegFit.cc#L310

**Comment by slava77** [comment]
@jshlee 
is there a presentation/slides in muon POG that describe the changes?
please add to the PR description

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
Can the isRPC be renamed to isGEMRPC so that we don't generate a new type?

**Comment by slava77** [comment]
also, please add some discussion what was already done in
https://github.com/cms-sw/cmssw/pull/13308
and is now present here and what was changed.

**Comment by jshlee** [comment]
@slava77  - no yet for muon POG, we will give one soon, but due to the ECFA deadlines we wanted to go ahead with this first. I'm not sure what you mean by isRPC.

@kpedro88 - yes, will fix this soon.

**Comment by slava77** [comment]
On 6/9/16 8:06 AM, Jason Lee wrote:
> @slava77 <https://github.com/slava77> - no yet for muon POG, we will
> give one soon, but due to the ECFA deadlines we wanted to go ahead with
> this first. I'm not sure what you mean by isRPC.

RPC type technically covers and doesn't overlap in meaning with GEM
(an extra layer of measurement in the same station as CSC)

**Comment by slava77** [comment]
Muon POG should sign off on this code.
a message from the conveners should be enough.

**Comment by kpedro88** [comment]
@jshlee: you may need to make some of these parameters/tags part of an Era
```
----- Begin Fatal Exception 09-Jun-2016 19:50:35 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=MuonIdProducer label='hltMuons'
Exception Message:
MissingParameter: Parameter 'GEMSegmentCollectionLabel' not found.
----- End Fatal Exception -------------------------------------------------
```

**Comment by slava77** [comment]
@jshlee @kpedro88 
how do we test muon reco in phase2?
(IIRC the matrix workflows are at best up to RECO:tracking_only)
-w upgrade has things like 10400.0 , with pure customs configs and full RECO, but they are dead.

**Comment by kpedro88** [comment]
@slava77 - I don't have any central workflow for muon reco at the moment...

**Comment by slava77** [comment]
On 6/10/16 9:23 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> - I don't have any central
> workflow for muon reco at the moment...

Is there a non-central workflow that works?

There was a premise of ECFA for this to go ahead sooner than
presentation/discussion in the POG.
but if the code can't run centrally, what's the point?

**Comment by kpedro88** [comment]
I've been trying to work on a "full reco" workflow (using Run2 calorimeters) this week. Unfortunately, changes in the Phase2 tracking objects are causing a lot of difficulties. But the muon systems haven't given me any trouble during my tests, at least.

I'm not sure what else can be done right now. If @jshlee and @calabria have private testing configurations, hopefully they can post them.

**Comment by slava77** [comment]
On 6/10/16 9:27 AM, Kevin Pedro wrote:
> I've been trying to work on a "full reco" workflow (using Run2
> calorimeters) this week. Unfortunately, changes in the Phase2 tracking
> objects are causing a lot of difficulties. But the muon systems haven't
> given me any trouble during my tests, at least.
> 
> I'm not sure what else can be done right now. If @jshlee
> <https://github.com/jshlee> and @calabria <https://github.com/calabria>
> have private testing configurations, hopefully they can post them.

anything cmsDriver-able would do, since this can be used in sample
production requests.

**Comment by jshlee** [comment]
@slava77 and @kpedro88 
I've been testing muon reco with the cmsDriver command below
cmsDriver.py SingleMuPt100_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW,RAW2DIGI,L1Reco,RECO --conditions auto:run2_mc --magField 38T_PostLS1 --datatier GEN-SIM --geometry Extended2015MuonGEMDev,Extended2015MuonGEMDevReco --eventcontent FEVTDEBUGHLT --era phase2_muon,run3_GEM -n 100 --fileout out_reco.root 

For the MissingParameter in hltMuons, should I use fillDescriptions or just update the hlt python files?

**Comment by slava77** [comment]
On 6/11/16 5:49 AM, Jason Lee wrote:
> @slava77 <https://github.com/slava77> and @kpedro88
> <https://github.com/kpedro88>
> I've been testing muon reco with the cmsDriver command below
> cmsDriver.py SingleMuPt100_cfi -s
> GEN,SIM,DIGI,L1,DIGI2RAW,RAW2DIGI,L1Reco,RECO --conditions auto:run2_mc
> --magField 38T_PostLS1 --datatier GEN-SIM --geometry
> Extended2015MuonGEMDev,Extended2015MuonGEMDevReco --eventcontent
> FEVTDEBUGHLT --era phase2_muon,run3_GEM -n 100 --fileout out_reco.root
> 
> For the MissingParameter in hltMuons, should I use fillDescriptions or
> just update the hlt python files?

fillDescriptions is preferred

more complicated situations are dealt with in
HLTrigger/Configuration/python/customizeHLTforCMSSW.py

**Comment by jshlee** [comment]
@kpedro88 - I tried to use fillDescriptions, but it doesn't seem to get pick up properly. I need to update the TrackAssociatorParameters 3 times.
Also, when this is run on hlt, all the inputs are from the hlt collections. So for the gem and me0 segments, we should make hltGEMSegments and hltME0Segments.

**Comment by kpedro88** [comment]
@jshlee it's hard for me to guess what the problem might be with so little information...

If the HLT collections are necessary at this point, then yes, we should make them. But for the moment, at least we can see a "missing collection" error instead of a "missing parameter" error.

**Comment by jshlee** [comment]
@kpedro88 - currently the problem is that the TrackAssociatorParameters should be derived from the TrackingTools.TrackAssociator.default_cfi, where GEMSegmentCollectionLabel are defined.

But, HLT does not import TrackAssociatorParameters from TrackingTools.TrackAssociator.default_cfi. 
The HLT defines everything in one python config file like OnLine_HLT_GRun.py.

Eventually, we would like to run this at the HLT as well.

**Comment by kpedro88** [comment]
@jshlee - I still don't fully understand the problem, but it sounds like you might have to make use of:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
maybe @slava77 can offer some additional advice.

**Comment by slava77** [comment]
I can't add more than what I mentioned in 
https://github.com/cms-sw/cmssw/pull/14833#issuecomment-225361963

**Comment by jshlee** [comment]
@slava77 and @kpedro88 - I tried the following
` edm::ParameterSetDescription desc;  
  edm::ParameterSetDescription descNested;
  descNested.add<edm::InputTag>("GEMSegmentCollectionLabel",edm::InputTag("gemSegments"));
  descNested.add<edm::InputTag>("ME0SegmentCollectionLabel",edm::InputTag("me0Segments"));
  descNested.setAllowAnything();
  desc.add<edm::ParameterSetDescription>("TrackAssociatorParameters", descNested);  
  desc.setAllowAnything();
  descriptions.addDefault(desc); `
but it doesn't seem to pick GEMSegmentCollectionLabel up.
any ideas on how I can fix this?

**Comment by kpedro88** [comment]
@jshlee - that looks correct to me... I'll try to think about why it might be failing, but I'm not an expert on how the HLT configs get generated.

**Comment by slava77** [comment]
@jshlee 
``` but it doesn't seem to pick GEMSegmentCollectionLabel up. ```
what did you mean by that?

**Comment by jshlee** [comment]
@slava77 - it still complains that GEMSegmentCollectionLabel is missing from hltMuons

**Comment by jshlee** [comment]
@slava77 and @kpedro88 - I solved the issue... there were more nested copies of `TrackAssociatorParameters`.

**Comment by jshlee** [comment]
@kpedro88 - I've updated this PR with the fillDescriptions. Can you test please.

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by makortel** [comment]
(sorry to jump in without knowing too much of the details)

The last error appears to be caused by the fact that `fillDescriptions()` were added to ESProduct (`MuonDetIdAssociator`), not to ESProducer (`DetIdAssociatorESProducer`). Unfortunately the `DetIdAssociatorESProducer` uses the "helper plugins via factory" pattern, for which there is currently no easy/straightforward solution for `fillDescriptions()` (AFAIK; we have plenty of those in tracking code). At least for this case you're probably better off with https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/HLTrigger/Configuration/python/customizeHLTforCMSSW.py.

**Comment by jshlee** [comment]
thanks @makortel! I'll think of another way to fix this.

**Comment by jshlee** [comment]
Since `fillDescriptions()` isn't an option, I used `try` to find gem and me0 detectors. 
Let me know if `customizeHLTforCMSSW.py` is preferred.

**Comment by slava77** [comment]
try-catch is a bad solution in this context.

**Comment by jshlee** [comment]
@slava77 - I changed to `customizeHLTforCMSSW.py`

**Comment by fwyzard** [comment]
-1

**Comment by fwyzard** [comment]
- entries in `customizeHLTforCMSSW.py` should be named after a specific pull request number, for ease of identification
  - the calls to the customisation functions should be dependent on the CMSSW version, to avoid problems when we backport the same file from 8.1.x to 8.0.x

**Comment by fwyzard** [comment]
A note for the framework export: if `fillDescriptions()` is currently not an option for modules use the "helper plugins via factory" pattern, the plugin mechanism should be extended to support it.

**Comment by slava77** [comment]
there are merge conflicts here
please fix

**Comment by jshlee** [comment]
fixed conflicts

**Comment by slava77** [comment]
@jshlee please confirm that ```runTheMatrix.py -l limited ``` runs

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@battibass, @HuguesBrun
Please confirm that these code changes were discussed within muon POG and that they are good to go from the muon POG point of view.
As of the PR submission, this wasn't yet presented in a meeting.

File: DataFormats/MuonReco/interface/Muon.h [hunk 1]
```diff
@@ -179,15 +179,20 @@ namespace reco {
...
     ///
-    /// number of chambers (MuonChamberMatches include RPC rolls)
+    /// number of chambers (MuonChamberMatches include RPC rolls, GEM and ME0 segments)
     int numberOfChambers() const { return muMatches_.size(); }
     /// number of chambers not including RPC matches (MuonChamberMatches include RPC rolls)
     int numberOfChambersNoRPC() const;
+    /// number of chambers not including GEM matches (MuonChamberMatches include RPC rolls, and also GEM)
+    int numberOfChambersNoGEM() const;    
+    /// number of chambers not including ME0 matches (MuonChamberMatches include RPC rolls, and also ME0)
+    int numberOfChambersNoME0() const;
```
Comment (slava77) [inline DataFormats/MuonReco/interface/Muon.h]: the purpose of numberOfChambersNoRPC was to capture the number of CSC-or-DT chambers.
I picked the name in 2011 not thinking of the future upgrades for a somewhat obvious reason.

numberOfChambersNoGEM and numberOfChambersNoME0 don't fit this purpose.
To make a more meaningful update, rename numberOfChambersNoRPC to be numberOfChambersCSCorDT.

File: DataFormats/MuonReco/src/Muon.cc [hunk 2]
```diff
@@ -74,6 +100,14 @@ int Muon::numberOfMatches( ArbitrationType type ) const
          matches++;
          continue;
       }
+      if(type == ME0SegmentAndTrackArbitration) {
+	if (chamberMatch->id.subdetId() == MuonSubdetId::ME0) matches += chamberMatch->segmentMatches.size();
+	continue;
+      }
+      if(type == GEMSegmentAndTrackArbitration) {
+	if (chamberMatch->id.subdetId() == MuonSubdetId::GEM) matches += chamberMatch->segmentMatches.size();
```
Comment (slava77) [inline DataFormats/MuonReco/src/Muon.cc]: this is a bogus arbitration.
Do you really want to keep it this way?

File: RecoMuon/DetLayers/src/MuonME0DetLayerGeometryBuilder.cc [hunk 3]
```diff
@@ -32,7 +32,7 @@ MuonME0DetLayerGeometryBuilder::buildEndcapLayers(const ME0Geometry& geo) {
   for (int endcap = -1; endcap<=1; endcap+=2) {
     int iendcap = (endcap==1) ? 0 : 1; // +1: forward, -1: backward
 
-    for(int layer = ME0DetId::minLayerId; layer <= ME0DetId::maxLayerId; ++layer) { 
+    for(int layer = ME0DetId::minLayerId+1; layer <= ME0DetId::maxLayerId; ++layer) {
```
Comment (slava77) [inline RecoMuon/DetLayers/src/MuonME0DetLayerGeometryBuilder.cc]: please clarify what this fixes

File: RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc [hunk 4]
```diff
@@ -752,6 +761,18 @@ bool MuonIdProducer::isGoodRPCMuon( const reco::Muon& muon )
...
 
+bool MuonIdProducer::isGEMMuon( const reco::Muon& muon )
+{
+  if(muon.track()->pt() < minPt_ || muon.track()->p() < minP_) return false;
+  return ( muon.numberOfMatches( reco::Muon::GEMSegmentAndTrackArbitration ) >= 1 );    
+}
+
+bool MuonIdProducer::isME0Muon( const reco::Muon& muon )
+{
+  if(muon.track()->pt() < minPt_ || muon.track()->p() < minP_) return false;
```
Comment (slava77) [inline RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc]: do you really want to apply the same pt cut for ME0 as used for CSC?
This value was selected/tuned to accept all muons within CSC acceptance.
ME0 goes beyond that.
pt of 0.5 GeV is probably low enough, for eta=3 this corresponds to p of 5 GeV. 
This is probably about 2 GeV above the rangeout.

File: TrackingTools/TrackAssociator/python/DetIdAssociatorESProducer_cff.py [hunk 5]
```diff
@@ -36,12 +36,18 @@
...
+    includeBadChambers = cms.bool(False),
+    includeGEM = cms.bool(False),
+    includeME0 = cms.bool(False)
 )
 
 # If running in Run 2, include bad chambers
 eras.run2_common.toModify( muonDetIdAssociator, includeBadChambers = True )
 
+# include GEM & ME0 for phase2
+eras.phase2_muon.toModify( muonDetIdAssociator, includeGEM = True )
```
Comment (slava77) [inline TrackingTools/TrackAssociator/python/DetIdAssociatorESProducer_cff.py]: this should be also in run3_GEM (if you add to run3_GEM, the change in phase2_muon is not necessary)

File: TrackingTools/TrackAssociator/src/TrackAssociatorParameters.cc [hunk 6]
```diff
@@ -63,6 +65,8 @@ void TrackAssociatorParameters::loadParameters( const edm::ParameterSet& iConfig
    if (useMuon) {
      dtSegmentsToken=iC.consumes<DTRecSegment4DCollection>(theDTRecSegment4DCollectionLabel);
      cscSegmentsToken=iC.consumes<CSCSegmentCollection>(theCSCSegmentCollectionLabel);
+     gemSegmentsToken=iC.consumes<GEMSegmentCollection>(theGEMSegmentCollectionLabel);
+     me0SegmentsToken=iC.consumes<ME0SegmentCollection>(theME0SegmentCollectionLabel);
```
Comment (slava77) [inline TrackingTools/TrackAssociator/src/TrackAssociatorParameters.cc]: please add these to consumes only when these detectors are enabled

File: TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc [hunk 7]
```diff
@@ -750,6 +755,11 @@ void TrackDetectorAssociator::fillMuon( const edm::Event& iEvent,
    if (! cscSegments.isValid()) 
      throw cms::Exception("FatalError") << "Unable to find CSCSegmentCollection in event!\n";
 
+   edm::Handle<GEMSegmentCollection> gemSegments;
+   iEvent.getByToken(parameters.gemSegmentsToken, gemSegments );
+   edm::Handle<ME0SegmentCollection> me0Segments;
+   iEvent.getByToken(parameters.me0SegmentsToken, me0Segments );
```
Comment (slava77) [inline TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc]: better call these only if these detectors logic is enabled in the configuration.

File: DataFormats/MuonReco/src/Muon.cc (hunk 2 above)
Comment (jshlee) [inline DataFormats/MuonReco/src/Muon.cc]: This is just a placeholder until the me0 and gem performance has been studied. For now we just need number of ME0 or GEM matches.

File: RecoMuon/DetLayers/src/MuonME0DetLayerGeometryBuilder.cc (hunk 3 above)
Comment (jshlee) [inline RecoMuon/DetLayers/src/MuonME0DetLayerGeometryBuilder.cc]: ME0 has 6 layers, minLayerId = 0 and maxLayerId = 6, so it was making one extra layer.

File: RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc (hunk 4 above)
Comment (jshlee) [inline RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc]: Yep, your right. I think the pt cut needs to be studied for both.
Shall I remove them for now and add in appropriate ones later?

**Comment by kpedro88** [comment]
@cmsbuild please test

File: RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc (hunk 4 above)
Comment (slava77) [inline RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc]: minP of 3 GeV is probably going to be OK here.
Just removing the selection will bump up the CPU time.
An "easy" option is to leave it as is, but add a comment that this should be reviewed once we get to physics studies.

Comment (jshlee) [inline RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc]: the values of minPt = cms.double(0.5) and minP = cms.double(2.5).
should I just use the minP for me0?

Comment (slava77) [inline RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc]: sounds reasonable.

File: DataFormats/MuonReco/src/classes_def.xml [hunk 8]
```diff
@@ -47,8 +47,8 @@
    <version ClassVersion="11" checksum="541727491"/>
    <version ClassVersion="10" checksum="4050071853"/>
   </class>
-  <class name="reco::MuonSegmentMatch" ClassVersion="10">
-   <version ClassVersion="10" checksum="754193003"/>
+  <class name="reco::MuonSegmentMatch" ClassVersion="11">
+   <version ClassVersion="11" checksum="2861296960"/>
```
Comment (slava77) [inline DataFormats/MuonReco/src/classes_def.xml]: old version-checksum line for version 10 should stay

File: RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc [hunk 9]
```diff
@@ -752,6 +761,20 @@ bool MuonIdProducer::isGoodRPCMuon( const reco::Muon& muon )
   return ( muon.numberOfMatchedRPCLayers( reco::Muon::RPCHitAndTrackArbitration ) > minNumberOfMatches_ );
 }
 
+bool MuonIdProducer::isGEMMuon( const reco::Muon& muon )
+{
+  // need to update min cuts on pt
```
Comment (slava77) [inline RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc]: GEM is in the same acceptance as CSC.
So, the same p/pt cuts should still  work.

Comment (jshlee) [inline RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc]: i use both pt and p for GEM

Comment (slava77) [inline RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc]: My comment was about the comment line ```// need to update min cuts on pt ```

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jshlee** [comment]
@slava77 and @kpedro88, can you retest please. The errors reported were from disk read failure.

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@jshlee @battibass @calabria @HuguesBrun 
what do we expect from this code: should the ME0 and GEM segments be counted in the definition of isTrackerMuon?
or was the intent to simply add the GEM and ME0 segment match information without modifying the isTrackerMuon?

e.g. in FourMuFlat wf 11200.0 (I'm running on top of #14865 )
all_sign728vssign727_fourmuflat2023wf11200p0c_recomuons_muons__reco_obj_eta2
we are now getting extra tracker muons
The number of PF muons goes up as well
all_sign728vssign727_fourmuflat2023wf11200p0c_recopfcandidates_particleflow__reco_obj_eta32

If I'm not mistaken, the way these muons make it to PF is by having numberOfMatches >= 2 (all other selections used in PFMuonAlgo based on TM\* selectors should be blind to GEM or ME0). In the setup on top of the current #14865  the ME0 is probably broken at the geometry level, so, I think that the increased counts of muons are from GE2/1 (more likely than from GE1/1).

**Comment by jshlee** [comment]
@slava @calabria - I understood that one of the goals is to have an Additional layer of “matched muons” using all the available segments in the muon system (DT, CSC, GEM) https://twiki.cern.ch/twiki/bin/viewauth/CMS/CMSMuonUpgradeSimulation. 
The way it is in this PR, the trackerMuon becomes this. Or else, I can add in a trackerMuon+ and keep trackerMuon unchanged.

**Comment by calabria** [comment]
@jshlee the idea is to have GEMMuon, ME0Muon, and also a new layer of matched muons where we use all the available segments. But as we agreed with @battibass @HuguesBrun and @trocino it is better to leave the tracker muon unchanged and another trackerMuon+

**Comment by slava77** [comment]
The easiest here would be to move the phase2 detectors to phase2Matches_ or something with a similar name vector of MuonChamberMatches. In this case the current selection methods will not pick them up.
The alternative is to go through all of the Muon and MuonSelectors code and make sure that functions apply only to DT and CSC detectors by default and can do something else.
From the DataFormats evolution point of view, this latter solution might be better though, but the selectors code is not the easiest to edit.

**Comment by trocino** [comment]
Yes, I agree, the trackerMuons should stay unchanged. So "isGoodTrackerMuon" should use the newly introduced "numberOfChambersCSCorDT" function, I guess. 

In theory, this problem should affect only the non-arbitrated trackerMuons, because the arbitration algorithms check only CSC and DT segments (apart from specific RPC or GEM arbitrations). So, in principle, all the practical use-cases should already be safe in the current version. 
But then I don't understand why the PFMuons are affected, I thought that only arbitrated trackerMuons could become PFMuons. Can you make sure that arbitrated trackerMuons are unchanged with this PR, @jshlee ?

**Comment by slava77** [comment]
On 6/21/16 7:47 AM, Daniele Trocino wrote:
> Yes, I agree, the trackerMuons should stay unchanged. So
> "isGoodTrackerMuon" should use the newly introduced
> "numberOfChambersCSCorDT" function, I guess.
> 
> In theory, this problem should affect only the non-arbitrated
> trackerMuons, because the arbitration algorithms check only CSC and DT
> segments (apart from specific RPC or GEM arbitrations). So, in
> principle, all the practical use-cases should already be safe in the
> current version.
> But then I don't understand why the PFMuons are affected, I thought that
> only arbitrated trackerMuons could become PFMuons. Can you make sure
> that arbitrated trackerMuons are unchanged with this PR, @jshlee
> <https://github.com/jshlee> ?


There are no checks for DT or CSC when arbitrated matches are counted
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_0_pre7/DataFormats/MuonReco/src/Muon.cc#L86
Segment masks are be filled irrespective of the detector
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_0_pre7/RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc#L941
well, there are some checks, but they don't capture all cases

**Comment by trocino** [comment]
You are right Slava, I got confused. It's the muon selector that uses only CSC and DT segments (because of the hard-coded range of this loop [1]), not the arbitration. 
In particular, "numberOfMatches" is affected by GEM and ME0 segments, but "numberOfMatchedStations" is not, for the same reason [2]. 

Concerning the main use-cases, this is the summary (if I'm not not much mistaken): 
- Tight, Soft and HighPt muons are not affected by GEM and ME0, because they either use "numberOfMatchedStations" or "isGoodMuon"; 
- Loose and Medium muons, and in general all PFMuons without further ID, are affected by the presence of GEM and ME0 matches. 

Would it be easier to make new arbitration types (in "numberOfMatches") that use only CSC and DT? Like "NoArbitrationCSCandDTonly" or so. This one in part. could be used inside MuonIdProducer to flag TrackerMuons. 
This change would not be transparent though. Apart from muon identification, "numberOfMatches" may be used also in other parts of the code, like validation and DQM, or even in private code.  

[1] https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DataFormats/MuonReco/src/MuonSelectors.cc#L338 
[2] https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_0_pre7/DataFormats/MuonReco/src/Muon.cc#L116

**Comment by jshlee** [comment]
@slava77 and @trocino - How about 
`numberOfMatches( ArbitrationType type, bool isUpgrade=false)` and `numberOfMatchedStations( ArbitrationType type = SegmentAndTrackArbitration, bool isUpgrade=false)` and for the non isUpgrade case, filter out gem and me0 hits.

**Comment by slava77** [comment]
On 6/21/16 9:27 AM, Jason Lee wrote:
> @slava77 <https://github.com/slava77> and @trocino
> <https://github.com/trocino> - How about
> |numberOfMatches( ArbitrationType type, bool isUpgrade=false)| and
> |numberOfMatchedStations( ArbitrationType type =
> SegmentAndTrackArbitration, bool isUpgrade=false)| and for the non
> isUpgrade case, filter out gem and me0 hits.
> 

Some mask to enable subdetectors may be more efficient.

I think that eventually all methods dealing with matches and arbitration
should become aware of the upgrade.
Long term this is easier with this default parameter
instead of introducing new default interface methods with CSCorDT and such.
Some refactoring may be necessary to not have the implementation blow up
in code line count.

**Comment by Martin-Grunewald** [comment]
+1

**Comment by slava77** [comment]
@jshlee 
please update on the progress here.
From the last messages exchanged, there should be changes made in this PR.

**Comment by jshlee** [comment]
@slava77 - we had a talk with the POG about this and are waiting for feedback from the muon reco contacts.

**Comment by jshlee** [comment]
@slava77 - The conclusion was to separate the identification.
Since all the arbitration depends on MuonChamberMatch.segmentMatches, gem and me0 are no longer saved here as before and saved instead in new MuonChamberMatch.gemMatches and MuonChamberMatch.me0Matches.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
For a change, the error is not from DAS
workflow 10624 in step3
```
   [1] Running path 'reconstruction_step'
   [2] Calling event method for module MuonProducer/'muons'
   [3] Calling produce method for unscheduled module CosmicsMuonIdProducer/'cosmicsVeto'
Exception Message:
BadRefCore RefCore: Request to resolve a null or invalid reference to a product of type 'std::vector<reco::Track>' has been detected.
```
this is very likely due to this PR.

**Comment by jshlee** [comment]
@slava77 - I've fixed this and tested it with workflow 10624. Please test again.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by dmitrijus** [comment]
+1

File: DataFormats/MuonReco/interface/MuonSegmentMatch.h [hunk 10]
```diff
@@ -46,7 +48,8 @@ namespace reco {
 
          DTRecSegment4DRef  dtSegmentRef;
          CSCSegmentRef      cscSegmentRef;
-
+	 GEMSegmentRef      gemSegmentRef;
+	 ME0SegmentRef      me0SegmentRef;
```
Comment (slava77) [inline DataFormats/MuonReco/interface/MuonSegmentMatch.h]: this, together with the use case pattern is becoming more and more bloated.
``` std::vector<reco::MuonSegmentMatch> segmentMatches ``` can have a value only among dtSegmentRef or cscSegmentRef, while 
```std::vector<reco::MuonSegmentMatch> gemMatches``` can have a meaningful value only for gemSegmentRef
... and inside the MuonChamberMatch only one of these vectors has a meaning.

This is probably still OK considering the most frequent access patterns: actual segments are used mainly in detailed detector analysis (truth matching and alike).
The alternatives will likely lead to polymorphic generalized more effective storage, but, very likely, access will require casting back to specific classes.

**Comment by slava77** [comment]
@jshlee @calabria @trocino 
after the last updates we now have more muons in the muon collections. As expected, they are not modifying the trackerMuons set. In the (no pileup) QCD dijet sample pt 600to800 (wf 10626.0) the total number of muons about doubles (from 15 to 29).

The additional muons contribute also to the earlyMuons collection and as a result affect tracking
e.g. 10626.0
all_sign728vsorig-not2ch_qcd600to800in13tev2023wf10626p0c_recotracks_generaltracks__reco_obj_algo
Since this is only in in-out iteration, the effect is going to be rather small, but will still likely enhance low quality track reconstruction (fakes and not).
For this workfow, the effect is mainly in the GEM direction (I didn't notice changes in ME0 direction).

If you want to decouple the new tracker muon types more completely, the earlyMuons should be modified to not use GEM and ME0.

I don't think it matters much for performance, but I wanted to get a clarification what was the intended behavior (fully decoupled, just new muons, or integrated already in the muon-driven tracking iterations).

**Comment by jshlee** [comment]
@slava77 - The idea was to go fully integrated already in the muon-driven tracking iterations, but not for identification.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14833 	4690943
- code changes are in line with the description and the follow up review/discussion. Only phase2 (and not yet existing run3) workflows are affected.
- jenkins tests pass and show changes only in 2023 workflow
- local tests with more 2023 workflows show expected behavior: new muons appear with GEM or RPC types and contribute to inside-out tracking iteration, but not to anything else from the ```_muons_``` collection.  On-disk increase is more significant in jetty events, expected for the loose matching definitions.  On-disk increase of the reco::Muon size is negligible for non-phase-2 workflows (only empty vectors data in the new data members). CPU increase in mostly-real-muon samples is small; expected to be more significant with pileup or in high-pt jets (I'm guessing it to be less than a per cent in high-PU, even without HGCAL).

**Comment by kpedro88** [comment]
@davidlange6 please merge for pre9 if possible

**Pull request merged**